Add fsync calls to pageserver

patins opened the pull request:
Add fsync calls to ensure durable storage of the layer files and
metadata.

For #427

patins commented on pageserver/src/layered_repository.rs:
```diff
@@ -177,7 +177,7 @@ impl Repository for LayeredRepository {
             ancestor_lsn: start_lsn,
         };
         std::fs::create_dir_all(self.conf.timeline_path(&dst, &self.tenantid))?;
```
I think we might also need to worry about fsync-ing the parent directories if they're being created

hlinnaka replied:
Good point. It depends on the OS and mount options, but in principle Linux at least requires is.

That's easy to forget. Perhaps we should have a helper function like 'create_dir_all_and_fsync` and use that everywhere instead of `create_dir_all`.

hlinnaka replied:
In theory, it's even worse than that: https://www.postgresql.org/message-id/Pine.OSF.4.61.0410311246390.238375%40kosh.hut.fi. That was a long time ago, and we decided back then that that's just crazy, and you should use mount options to prevent that craziness. Modern Linux and filesystem versions don't have that problem AFAIK.

patins commented on pageserver/src/layered_repository.rs:
```diff
@@ -1359,6 +1392,14 @@ impl LayeredTimeline {
             layer.unload()?;
         }
 
+        drop(layers);
```
I don't see any obvious races that could be caused by this, but shout out if one comes to mind

hlinnaka commented on pageserver/src/layered_repository.rs:
```diff
@@ -355,13 +355,34 @@ impl LayeredRepository {
+        let timeline_path = conf.timeline_path(&timelineid, &tenantid);
+        let path = timeline_path.join("metadata");
+        // use OpenOptions to ensure file presence is consistent with first_save
+        let mut file = OpenOptions::new()
```
Truncating the file in place is scary. If you crash after the truncation, but before the new contents have been written out, you don't have any version of it.

In PostgreSQL, the 'control file' is somewhat similar to the metadata file here. It's created as a fixed-size file when the cluster is initialized, and it's overwritten on every checkpoint. It's small, less than 512 bytes, and we assume that it's written atomically. We could do the same here, but that requires that the file's size is never changed.

An alternative is to create the new metadata file with a temporary name, and rename() it in place.

patins replied:
Is the rename crash safe?
Also would this be safe if I got rid of the truncate and always wrote a 512 byte vec (zeroes at the end)?

hlinnaka replied:
> Is the rename crash safe?
I think the safe pattern is
1. create with temp name
2. fsync the file. (Or at least the contents, i.e. fdatasync or File::sync_data).
3. rename the file in place
4. fsync the file metadata (File::sync_all)

> Also would this be safe if I got rid of the truncate and always wrote a 512 byte vec (zeroes at the end)?

Yeah, that's acceptable. I don't think that guarantee is written down in any authoritative source like POSIX, but we assume that in Postgres with the control file and haven't had problems, and other software make similar assumptions.

SomeoneToIgnore commented on pageserver/src/layered_repository.rs:
```diff
@@ -355,13 +355,34 @@ impl LayeredRepository {
+        assert!(metadata_bytes.len() <= MAX_SAFE_SIZE);
+
+        if file.write(&metadata_bytes)? != metadata_bytes.len() {
+            bail!("Could not write all the metadata bytes in a single call");
```
How would things work after this `bail`? It it so, that there will be a file left with the `metadata` name, but incorrect contents?

Sounds a bit dangerous IMO if really the case.

patins replied:
Yeah, I'm not sure about that case. I copied postgres semantics here:
https://github.com/postgres/postgres/blob/c30f54ad732ca5c8762bb68bbe0f51de9137dd72/src/common/controldata_utils.c#L212

Makes me wonder what happens if we happen to load a debugger just as this write is happening

SomeoneToIgnore replied:
Maybe I'm not aware of something, but the metadata file is our idea and we control it entirely: I would've written that file "normally", entirely or not create it at all (retrying later?)

But never mind me much, not insisting you should do that, just expressing the confusion.

hlinnaka replied:
Yeah, we have full control of the metadata file, it's our invention. But it's similar to Postgres' control file, so it makes sense to look there for inspiration.

patins commented:
Need to also take a closer look at how we load layer files

patins commented on pageserver/src/create_dir.rs:
```diff
@@ -0,0 +1,39 @@
+use std::{fs::File, io::ErrorKind, path::Path};
```
These utilities don't do what I want at the moment. Will fix and document

patins commented on pageserver/src/layered_repository.rs:
```diff
@@ -352,13 +354,35 @@ impl LayeredRepository {
+        const MAX_SAFE_SIZE: usize = 512;
+        assert!(metadata_bytes.len() <= MAX_SAFE_SIZE);
+
+        metadata_bytes.resize(MAX_SAFE_SIZE, 0u8);
```
I reworked this approach so it always writes the same # of bytes and doesn't truncate. Happy to switch to rename approach as well, let me know

patins commented:
This PR was getting kind of big, so I split into individual commits. I added functionality for metadata checksums & rejecting future layer files + added tests for both of these features.

SomeoneToIgnore reviewed (approved):
Looks like I need to adjust the `RelishStorage` impls similarly, thanks for showing the way 🙂
